src/scripts/tag-and-release.ts

```diff
@@ -0,0 +1,103 @@
+
+const execSyncWithErrorHandling = (command: string) => {
+  // eslint-disable-next-line functional/no-try-statements
+  try {
+    return execSync(command).toString();
+  } catch (error: any) {
+    console.info(error.message);
+    console.info('STDOUT', error.stdout.toString());
+    console.info('STDERR', error.stderr.toString());
+    process.exit(1);
+  }
+};
+
+export const createGithubRelease = async (repo: string, tagName: `v${string}`) => {
+  if (!process.env.GH_ACCESS_TOKEN) {
+    console.info(`GH_ACCESS_TOKEN not set. Skipping release creation`);
+    return null;
+  }
+  // Ensure the GH_ACCESS_TOKEN secret is set on Github and has the relevant permissions
+  const octokit = new Octokit({ auth: process.env.GH_ACCESS_TOKEN });
+  const createRelease = async () =>
+    octokit.rest.repos.createRelease({
+      owner: 'api3dao',
+      repo,
+      tag_name: tagName, // eslint-disable-line camelcase
+      generate_release_notes: true, // eslint-disable-line camelcase
+    });
+  console.info(`Creating Github release...`);
+  const goRes = await go(createRelease, { totalTimeoutMs: 15_000 });
+  if (!goRes.success) {
+    // We don't want to fail CI if the release fails to create. This can be done manually through Github's UI
+    console.info(`Unable to create Github release`);
+    console.info(goRes.error.message);
+    return null;
+  }
+  return goRes.data;
+};
+
+export const main = async () => {
+  const repo = process.argv[2];
+  if (!repo) throw new Error(`First argument must be the repo name`);
+
+  console.info('Ensuring working directory is clean...');
+  const gitStatus = execSyncWithErrorHandling('git status --porcelain');
+  if (gitStatus !== '') throw new Error('Working directory is not clean');
+
+  console.info('Ensuring we are on the main branch...');
+  const branch = execSyncWithErrorHandling('git branch --show-current');
+  if (branch !== 'main\n') throw new Error('Not on the main branch');
+
+  console.info('Ensuring we are up to date with the remote...');
+  execSyncWithErrorHandling('git fetch');
+
+  const gitDiff = execSyncWithErrorHandling('git diff origin/main');
+  if (gitDiff !== '') throw new Error('Not up to date with the remote');
+
+  const packageJson = JSON.parse(readFileSync(join(__dirname, '../package.json'), 'utf8')) as any;
+  const { version } = packageJson;
+  console.info(`Version set to ${version}...`);
+
+  const gitTag = execSyncWithErrorHandling(`git tag -l '*v${version}*'`);
+  if (gitTag !== '') throw new Error(`git tag v${version} already exists`);
+
+  console.info('Creating new annotated git tag...');
+  execSyncWithErrorHandling(`git tag -a v${version} -m "v${version}"`);
+
+  console.info('Pushing git tag...');
+  // NOTE: in order to push, a valid access token is expected as GH_ACCESS_TOKEN
+  execSyncWithErrorHandling(`git push origin v${version} --no-verify`);
+
+  await createGithubRelease(repo, `v${version}`);
+
+  console.info(`Done!`);
+};
+
+main()
```

Comment:
It's not good practice to export the module with side effect (executing the release). I think it should only export the feature and let the consumer call it.

Reply:
This was also intentional as importing projects would all need to add their own `scripts` directory and copy-paste some boilerplate. I also don't feel strongly here though

Reply:
I think that pasting some boilerplate is inevitable. You need to edit the `.yml` file for the CI and define a target in scripts. At which point you already can have the `scripts` folder and be explicit in calling the release function from commons.